File: test-browser.js
/**
 * Browser Integration Test Script - Models Page Scan Test
 */

import { chromium } from "playwright";

async function testModelsScan() {
  console.log("=".repeat(60));
  console.log("PLAYWRIGHT TEST: Models Page Scan Functionality");
  console.log("=".repeat(60));

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  const errors = [];
  const logs = [];

  // Capture console messages
  page.on("console", (msg) => {
    const text = msg.text();
    logs.push({ type: msg.type(), text });
    // Only log debug messages
    if (msg.type() === "log" || msg.type() === "error") {
      console.log(`[CONSOLE ${msg.type()}] ${text}`);
    }
  });

  // Capture page errors
  page.on("pageerror", (error) => {
    console.error("[PAGE ERROR]", error.message);
    errors.push(error.message);
  });

  try {
    // Step 1: Navigate to models page
    console.log("\n[STEP 1] Navigating to Models page...");
    await page.goto("http://localhost:3000/models", {
      waitUntil: "networkidle",
      timeout: 30000,
    });
    console.log("[STEP 1] Page loaded");

    // Wait for Socket.IO connection
    await page.waitForTimeout(2000);

    // Check if models table exists
    const tableExists = await page.$(".models-table");
    console.log("[STEP 1] Models table exists:", !!tableExists);

    // Count initial rows
    const initialRows = await page.$$eval(".models-table tbody tr", (rows) => rows.length);
    console.log("[STEP 1] Initial row count:", initialRows);

    // Step 2: Click scan button
    console.log("\n[STEP 2] Clicking 'Scan Filesystem' button...");
    const scanButton = await page.$('[data-action="scan"]');
    if (!scanButton) {
      throw new Error("Scan button not found!");
    }
    await scanButton.click();
    console.log("[STEP 2] Scan button clicked");

    // Step 3: Wait for scan to complete
    console.log("\n[STEP 3] Waiting for scan to complete...");
    await page.waitForTimeout(5000);

    // Check notifications
    const notifications = await page.$$("#notifications > div");
    console.log("[STEP 3] Notifications count:", notifications.length);
    for (const notif of notifications) {
      const text = await notif.textContent();
      console.log("[STEP 3] Notification:", text);
    }

    // Step 4: Check for updated data
    console.log("\n[STEP 4] Checking for updated data...");
    const finalRows = await page.$$eval(".models-table tbody tr", (rows) => rows.length);
    console.log("[STEP 4] Final row count:", finalRows);

    // Get first row data
    if (finalRows > 0) {
      const cells = await page.$$eval(".models-table tbody tr:first-child td", (cells) =>
        cells.map((c) => c.textContent?.trim())
      );
      console.log("[STEP 4] First model row:");
      console.log("  Name:", cells[0]);
      console.log("  Status:", cells[1]);
      console.log("  Arch:", cells[2]);
      console.log("  Params:", cells[3]);
      console.log("  Quant:", cells[4]);
      console.log("  Ctx:", cells[5]);
      console.log("  Embed:", cells[6]);
      console.log("  Blocks:", cells[7]);
      console.log("  Heads:", cells[8]);
      console.log("  Size:", cells[9]);
    }

    // Summary
    console.log("\n" + "=".repeat(60));
    console.log("TEST SUMMARY");
    console.log("=".repeat(60));
    console.log(`Initial rows: ${initialRows}`);
    console.log(`Final rows: ${finalRows}`);
    console.log(`Errors: ${errors.length}`);

    // Check for key success/failure indicators
    const scanSuccess = logs.some((l) => l.text.includes("SUCCESS") && l.text.includes("scanModels"));
    const scanError = logs.some((l) => l.text.includes("ERROR") && l.text.includes("scanModels"));
    const modelsRefreshed = logs.some((l) => l.text.includes("refreshModels") && l.text.includes("complete"));

    console.log("\nKey indicators:");
    console.log(`  Scan success: ${scanSuccess}`);
    console.log(`  Scan error: ${scanError}`);
    console.log(`  Models refreshed: ${modelsRefreshed}`);

    if (errors.length > 0) {
      console.log("\nErrors:");
      errors.forEach((err, i) => console.log(`  ${i + 1}. ${err}`));
    }

    console.log("\n" + (errors.length === 0 && scanSuccess ? "✓ TEST PASSED" : "✗ TEST NEEDS INVESTIGATION"));

  } catch (error) {
    console.error("\n[TEST ERROR]", error.message);
    errors.push(error.message);
  } finally {
    await browser.close();
    console.log("\n" + "=".repeat(60));
    console.log("TEST COMPLETE");
    console.log("=".repeat(60));
  }

  return errors.length === 0;
}

testModelsScan()
  .then((success) => {
    process.exit(success ? 0 : 1);
  })
  .catch((err) => {
    console.error("Fatal error:", err);
    process.exit(1);
  });

File: test-browser.py
#!/usr/bin/env python3
"""
Interactive browser test for Llama Proxy Dashboard
"""

from playwright.sync_api import sync_playwright
import sys


def test_full():
    results = []

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=False)  # Visible browser
        page = browser.new_page()

        # Capture all console messages
        console_output = []

        def handle_console(msg):
            console_output.append(f"[{msg.type}] {msg.text}")
            if msg.type == "error":
                results.append(f"ERROR: {msg.text}")

        page.on("console", handle_console)

        # Test 1: Load dashboard
        print("=" * 60)
        print("TEST 1: Loading Dashboard")
        print("=" * 60)
        page.goto("http://localhost:3000")
        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(2000)

        # Check page loaded
        title = page.title()
        print(f"Page title: {title}")

        # Check elements
        app = page.locator("#app")
        dashboard = page.locator(".dashboard-page")
        nav = page.locator("header")

        print(f"App container: {'✅' if app.count() > 0 else '❌'}")
        print(f"Dashboard: {'✅' if dashboard.count() > 0 else '❌'}")
        print(f"Header: {'✅' if nav.count() > 0 else '❌'}")

        if app.count() == 0:
            results.append("App container not found")

        # Test 2: Socket.IO connection
        print("\n" + "=" * 60)
        print("TEST 2: Socket.IO Connection")
        print("=" * 60)
        socket_works = page.evaluate("""() => {
            return typeof window.socketClient !== 'undefined' && 
                   window.socketClient.isConnected === true;
        }""")
        print(f"Socket.IO connected: {'✅' if socket_works else '❌'}")
        if not socket_works:
            results.append("Socket.IO not connected")

        # Test 3: Navigate to Models page
        print("\n" + "=" * 60)
        print("TEST 3: Navigate to Models Page")
        print("=" * 60)
        models_link = page.locator("a[href='/models']")
        if models_link.count() > 0:
            models_link.first.click()
            page.wait_for_load_state("networkidle")
            page.wait_for_timeout(2000)

            models_page = page.locator(".models-page")
            print(f"Models page loaded: {'✅' if models_page.count() > 0 else '❌'}")

            # Check if models loaded
            models_loaded = page.evaluate("""() => {
                const el = document.querySelector('.models-page');
                return el && el.innerText.length > 0;
            }""")
            print(f"Models content visible: {'✅' if models_loaded else '❌'}")
        else:
            print("Models link not found")
            results.append("Models navigation link not found")

        # Test 4: Navigate back to Dashboard
        print("\n" + "=" * 60)
        print("TEST 4: Navigate back to Dashboard")
        print("=" * 60)
        dashboard_link = page.locator("a[href='/']")
        if dashboard_link.count() > 0:
            dashboard_link.first.click()
            page.wait_for_load_state("networkidle")
            page.wait_for_timeout(2000)

            dashboard = page.locator(".dashboard-page")
            print(f"Back to Dashboard: {'✅' if dashboard.count() > 0 else '❌'}")
        else:
            print("Dashboard link not found")

        # Test 5: Check for router status card
        print("\n" + "=" * 60)
        print("TEST 5: Router Status Card")
        print("=" * 60)
        router_card = page.locator(".llama-router-status-card")
        print(f"Router card present: {'✅' if router_card.count() > 0 else '❌'}")

        # Check router status text
        status_text = page.locator(".badge-text").first.inner_text()
        print(f"Router status: {status_text}")

        # Test 6: Check all buttons work (no JS errors on click)
        print("\n" + "=" * 60)
        print("TEST 6: Button Interactions")
        print("=" * 60)
        buttons = page.locator("button").all()
        print(f"Total buttons on page: {len(buttons)}")

        # Try clicking a button (e.g., refresh if exists)
        refresh_btn = page.locator("[data-action='refresh']")
        if refresh_btn.count() > 0:
            refresh_btn.first.click()
            page.wait_for_timeout(1000)
            print("Refresh button clicked: ✅")
        else:
            print("Refresh button not found (may be normal)")

        # Take screenshot
        page.screenshot(path="/tmp/llama-dashboard-test.png", full_page=True)
        print("\n📸 Screenshot saved: /tmp/llama-dashboard-test.png")

        browser.close()

    # Final Results
    print("\n" + "=" * 60)
    print("FINAL RESULTS")
    print("=" * 60)

    if results:
        print(f"❌ {len(results)} error(s):")
        for r in results:
            print(f"  - {r}")
        return 1
    else:
        print("✅ All tests passed!")
        print("✅ Application working correctly with Socket.IO!")
        return 0


if __name__ == "__main__":
    sys.exit(test_full())
